src/test/expressions/expression_test_helpers.hpp

```diff
@@ -0,0 +1,150 @@
+T make_arg() {
+  return 0.4;
+}
+template <typename T, require_eigen_vector_t<T>* = nullptr>
+T make_arg() {
+  T res(1);
+  res << 0.1;
+  return res;
+}
+template <typename T, require_eigen_matrix_t<T>* = nullptr>
+T make_arg() {
+  T res(1, 1);
+  res << 0.1;
+  return res;
+}
+template <typename T, require_std_vector_t<T>* = nullptr>
+T make_arg() {
+  using V = value_type_t<T>;
+  V tmp = make_arg<V>();
+  T res;
+  res.push_back(tmp);
+  return res;
+}
+template <typename T, require_same_t<T, std::minstd_rand>* = nullptr>
+T make_arg() {
+  return std::minstd_rand(0);
+}
+
+template <typename T, require_arithmetic_t<T>* = nullptr>
+void expect_eq(T a, T b, const char* msg) {
+  EXPECT_EQ(a, b) << msg;
+}
+
+void expect_eq(math::var a, math::var b, const char* msg) {
+  EXPECT_EQ(a.val(), b.val()) << msg;
+}
+
+template <typename T, require_arithmetic_t<T>* = nullptr>
+void expect_eq(math::fvar<T> a, math::fvar<T> b, const char* msg) {
+  expect_eq(a.val(), b.val(), msg);
+  expect_eq(a.tangent(), b.tangent(), msg);
+}
+
+template <typename T1, typename T2, require_all_eigen_t<T1, T2>* = nullptr,
+          require_vt_same<T1, T2>* = nullptr>
+void expect_eq(const T1& a, const T2& b, const char* msg) {
+  EXPECT_EQ(a.rows(), b.rows()) << msg;
+  EXPECT_EQ(a.cols(), b.cols()) << msg;
+  const auto& a_ref = math::to_ref(a);
+  const auto& b_ref = math::to_ref(b);
+  for (int i = 0; i < a.rows(); i++) {
+    for (int j = 0; j < a.cols(); j++) {
+      expect_eq(a_ref(i, j), b_ref(i, j), msg);
+    }
+  }
+}
+
+template <typename T>
+void expect_eq(const std::vector<T>& a, const std::vector<T>& b,
+               const char* msg) {
+  EXPECT_EQ(a.size(), b.size());
+  for (int i = 0; i < a.size(); i++) {
+    expect_eq(a[i], b[i], msg);
+  }
+}
+
+template <typename T, require_not_st_var<T>* = nullptr>
+void expect_adj_eq(const T& a, const T& b, const char* msg) {}
+
+void expect_adj_eq(math::var a, math::var b, const char* msg) {
+  EXPECT_EQ(a.adj(), b.adj()) << msg;
+}
+
+template <typename T1, typename T2, require_all_eigen_t<T1, T2>* = nullptr,
+          require_vt_same<T1, T2>* = nullptr>
+void expect_adj_eq(const T1& a, const T2& b, const char* msg) {
+  EXPECT_EQ(a.rows(), b.rows()) << msg;
+  EXPECT_EQ(a.cols(), b.cols()) << msg;
+  const auto& a_ref = math::to_ref(a);
+  const auto& b_ref = math::to_ref(b);
+  for (int i = 0; i < a.rows(); i++) {
+    for (int j = 0; j < a.cols(); j++) {
+      expect_adj_eq(a_ref(i, j), b_ref(i, j), msg);
+    }
+  }
+}
+
+template <typename T>
+void expect_adj_eq(const std::vector<T>& a, const std::vector<T>& b,
+                   const char* msg) {
+  EXPECT_EQ(a.size(), b.size()) << msg;
+  for (int i = 0; i < a.size(); i++) {
+    expect_adj_eq(a[i], b[i], msg);
+  }
+}
+
+#define TO_STRING_(x) #x
+#define TO_STRING(x) TO_STRING_(x)
+#define EXPECT_STAN_EQ(a, b) \
+  stan::test::expect_eq(     \
+      a, b, "Error in file: " __FILE__ ", on line: " TO_STRING(__LINE__))
```

Comment:
I was thinking about trying to trigger this error to see if it was working.

We don't want to burn too much time testing the testing framework, but I think spot checks here to make sure:

1. A functions that doesn't take an expression fails

2. A functions that evals twice fails

3. A functions whose output is different for expressions and values fails

4. A function whose gradients are different for expressions and values fails

In that regard I think you could define a few functions like that in this header and put them in the banned function list. I think it would be convenient to make ```--only-functions``` ignore the banned functions list so we could unit test those things.

I'm not sure we can actually test #1 cause that's a compiler failure. So if we can't we can't and that's fine.

Reply:
Added tests (including 1). I expect setting up those on Jenkins would be quite a hassle, as they need to be run separately, 1 is expected to fail the compilation, 2 and 3 are expected to fail 3/3 tests and 4 is expected to fail 2/3 tests (prim passes).

> I think it would be convenient to make --only-functions ignore the banned functions list so we could unit test those things.

done

Reply:
Hmm, yeah I could see 1 being a problem.

But 2-4 are worth doing. Maybe we can limit the scope of these and get somewhere.

A way to mostly-do these tests is take the generated code for them and just add those as unit tests.

We're not testing the code gen that way, but we're roughly testing the check functions. I expect if code gen breaks we'll know. I do want a little assurance that the value/gradient/double-eval tests are working though.

Reply:
I wrapped those in python unittests so they can all be run on jenkins

Reply:
How long do the unit tests take to run? I tried it out and let it run for like 10 mins but I was getting no output or anything.

Reply:
/runTests.py -j8 src/test/expressions runs for 4 minutes on Jenkins.

Reply:
Oops I meant the tests of the tests.

```
python test_expression_testing_framework.py
```

Reply:
Those just need to run tests for 4 functions, so they should be fast.